crypto/rand/rand_vms.c

```diff
@@ -148,12 +465,19 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
     /*
      * Try not to overfeed the pool
      */
-    if (total_length > bytes_remaining)
+    if (total_length > bytes_remaining) {
         total_length = bytes_remaining;
+        total_length_64bit = 0;
+    } else if (total_length + total_length_64bit > bytes_remaining)
+        total_length_64bit = bytes_remaining - total_length;
 
     /* We give the pessimistic value for the amount of entropy */
-    rand_pool_add(pool, (PTR_T)data_buffer, total_length,
-                  total_length / ENTROPY_FACTOR);
+    rand_pool_add(pool, (unsigned char *)data_buffer, total_length,
+                  8 * total_length / ENTROPY_FACTOR);
```

Comment:
Is compiler allowed to optimize this as `(8 / ENTROPY_FACTOR) * total_length`? Which would yield 0 for any total_length... In other words I think explicit `(8 * total_length) / ENTROPY_FACTOR` would be appropriate here...

Reply:
I thought `*` and `/` were left-to-right operators. Are compilers allowed to move around operations like you suggest, even though that would give different results?

Reply:
The language spec mandates an order of operations, which IIRC is left-to-right, but I did not double-check.

Reply:
Note that it was actually a question. But here is another question. How universal is suggestion that * and / are left-to-right? Do you recognize that it actually depends on type?  For example contemporary compilers have no problem optimizing `8*x/4` as `2*x` for signed integer. But not unsigned. [Do you want to know why?] Though older gcc optimizes it as `2*x` even for unsigned. And don't get me started on floating point... Of course `8*x/4` is very different from `8*x/20` and compilers should recognize that. But can you be *sure*? For any of compilers that are out there? And think of it even like this. Does reader of the code have to keep all those details in mind? Actual value of ENTROPY_FACTOR, dependency on type, which type is involved in operation...

Reply:
> every compiler that are out there?

Well, of course one can say that it has to work with just one, VMS... Yeah, consider it a side note...

Reply:
Thanks for that heads up.  Me, I'm thinking that paranoia is fine, as long as it doesn't cost our sanity.  In this case, I'll leave it be, because the VMS compilers do fine with it, and considering VSI are keeping a close eye on OpenSSL, I wouldn't be surprised if it's one of the test beds for new ones.

Reply:
Fair enough. BTW, would a test fail if 0 is passed to rand_pool_add here?

Reply:
The question was if any of the tests executed in equivalent of 'make test' would fail. I mean question is *how* would the keep an eye on it? They obviously won't single-step whole code base in debugger...